### ISSUE TITLE
Missing values handling at prediction

### ISSUE THREAD

**Issue by ronencozen** [dmlc/xgboost#1607]
Hi ,

I am using xgboost package in R, and run into a problem when predicting a test set which has missing values in one or more features, the observations which has missing values are not predicted.
here is a reproducible example using the iris dataset.

```
library(dplyr)
library(caret)

# Data Prep
  iris2 <- iris %>%
    filter(Species %in% c('setosa' , 'virginica'))
  iris2$Species <- as.factor(as.character(iris2$Species))
  iris2$Petal.Width <- ifelse(iris2$Petal.Width == 0.2, NA, iris2$Petal.Width)

  set.seed(12)
  trainIndex <- createDataPartition(iris2$Species, p = .8, list = FALSE, times = 1)
  irisTrain <- iris2[ trainIndex,]
  irisTest  <- iris2[-trainIndex,]

# Training
library(xgboost)
  fitControl <- trainControl(
    method = "cv",
    number = 10,
    repeats = 1,
    summaryFunction = twoClassSummary,
    classProbs = TRUE,
    allowParallel = TRUE)

  xgbGrid <-  expand.grid(
    nrounds = c(2),
    max_depth = c(3),
    eta = c(0.1),
    gamma = 0,
    colsample_bytree = 1,
    min_child_weight = 1)

  xgbModel <- train(
    Species ~ .,
    data = irisTrain,
    preProc = c("center", "scale"),
    method = "xgbTree",
    metric = "ROC",
    trControl = fitControl,
    tuneGrid = xgbGrid,
    verbose = 2)

# Prediction
  pred <- predict(xgbModel, irisTest)
  confusionMatrix(pred, irisTest$Species)
```
### Error

> > confusionMatrix(pred, irisTest$Species)
> > Error in table(data, reference, dnn = dnn, ...) : 
> >   all arguments must have the same length

**Comment by khotilov** [comment]
Perhaps `predict(xgbModel, irisTest, na.action = na.pass)` ?
It seems like a caret's problem.

**Comment by ronencozen** [comment]
just run predict with parameter na.action = na.pass
and the error message is:

>  Error in xgb.DMatrix(as.matrix(newdata)) : 
  There are NAN in the matrix, however, you did not set missing=NAN 

do you think na.action is not implemented on the caret side ?

**Comment by khotilov** [comment]
You OP example runs fine for me with the latest caret and xgboost installed, and using `na.action = na.pass` in both train and predict.

That last error comes from xgboost and what it tells you is kind of obvious: your data has NaN's

**Comment by ronencozen** [comment]
### I followed your guidelines, but i am still facing an error:

>  There are NAN in the matrix, however, you did not set missing=NAN

### here is the session info:
> > sessionInfo()
R version 3.3.0 (2016-05-03)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] xgboost_0.4-4   caret_6.0-71    ggplot2_2.1.0   lattice_0.20-33 dplyr_0.4.3     plyr_1.8.3     

### modified train and predict functions to include na.action:
`xgbModel <- train(
    Species ~ .,
    data = irisTrain,
    preProc = c("center", "scale"),
    method = "xgbTree",
    metric = "ROC",
    trControl = fitControl,
    tuneGrid = xgbGrid,
    verbose = 2,
    na.action = na.pass)

pred <- predict(xgbModel, irisTest, na.action = na.pass)`

**Comment by khotilov** [comment]
It seems like your xgboost package was installed from cran, which is based on very old (about a year old code) that had default setting `missing = 0` in xgb.DMatrix:
https://github.com/cran/xgboost/blob/3a9b4849152ad856839da0c7fc32dfb0bf0c3d46/R/xgb.DMatrix.R#L21
caret doesn't provide any means to change this parameter:
https://github.com/topepo/caret/blob/d11c6f41d6e34bfd317ec49e8323a2042b6a470d/models/files/xgbTree.R#L53

See how to install the recent version: https://xgboost.readthedocs.io/en/latest/build.html#r-package-installation where the default is `missing = NA`

**Comment by ronencozen** [comment]
I installed the latest xgboost, and since  missing = NA is the default, i didn't include this parameter either in the train or predict commands, and received the following error:

> Error in na.fail.default(list(Species = c(1L, 1L, 1L, 1L, 1L, 1L, 1L,  : 
  missing values in object

**Comment by hadjipantelis** [comment]
@khotilov Thank you for your clarifications so far. Is this now addressed in the new CRAN version?
